sql-plugin/src/main/340+/scala/com/nvidia/spark/rapids/shims/CastingConfigShim.scala

```diff
@@ -18,11 +18,16 @@ package com.nvidia.spark.rapids.shims
 
 import org.apache.spark.sql.catalyst.expressions.{Cast, Expression}
 
-object CastIgnoreTimeZoneShim {
+object CastingConfigShim {
   /** Remove TimeZoneId for Cast if needsTimeZone return false. */
   def ignoreTimeZone(e: Expression): Expression = e match {
     case c: Cast if c.timeZoneId.nonEmpty && !c.needsTimeZone =>
       c.withTimeZone(null)
     case _ => e
   }
+
+  // Use public member of Cast class rather than reflection
+  def ansiEnabled(e: Expression): Boolean = {
```

Comment:
As a side not in 3.4.0 what are we doing for `EvalMode.TRY`?  It shows up as an AnsiCast when calling `ansiEnabled`, but it behaved differently. I have not dug into it that deeply. Do we need to add tests for try_mode? is having `ansiEnabled` in the shims the way it is right now going to make it difficult to add tests for try mode? Are we just not going to support try mode?

Reply:
I believe there is a myriad of other 3.4.0 related changes we need to test, investigate, and fix -- with many related to the combined casting approach new to 3.4.0. I think our current approach is to get it building/passing unit tests, then start fixing the inconsistencies in results as part of fixing integration tests.

Reply:
@revans2 - IIUC EvalMode.TRY is similar to `ansiMode`, except that TRY mode executes as if it would fail on errors but would capture the errors and return `null` as result. We currently don't support any TRY_* functions such as tryAdd, trySubtract etc. 

But I do a problem when TRY_CAST, TRY_ADD is used. For example: In the query, it shows up as `Cast` expression irrespective of whether it is `try_cast` or `cast` - [link here](https://github.com/apache/spark/blob/34d5272663ce4852ca5b2daa665983a321b42060/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala#L2451)

So when we use TRY_CAST with ansiEnabled=true, Plugin would run the regular Cast and throw error instead of nulls.
Basically TRY_CAST should not depend whether ansiEnabled is set or not. But in our case it would behave differently.

```
scala>  spark.sql("select * from boolsTable").show
+-----+
|bools|
+-----+
|  def|
|  abc|
+-----+

scala> spark.conf.set("spark.sql.ansi.enabled", "true")

scala>  spark.sql("select TRY_CAST(bools AS boolean) from boolsTable").show
22/11/10 03:11:49 WARN GpuOverrides: 
!Exec <CollectLimitExec> cannot run on GPU because the Exec CollectLimitExec has been disabled, and is disabled by default because Collect Limit replacement can be slower on the GPU, if huge number of rows in a batch it could help by limiting the number of rows transferred from GPU to CPU. Set spark.rapids.sql.exec.CollectLimitExec to true if you wish to enable it
  @Partitioning <SinglePartition$> could run on GPU
  *Exec <ProjectExec> will run on GPU
    *Expression <Alias> cast(try_cast(bools#107 as boolean) as string) AS bools#132 will run on GPU
      *Expression <Cast> cast(try_cast(bools#107 as boolean) as string) will run on GPU
        *Expression <Cast> try_cast(bools#107 as boolean) will run on GPU
    *Exec <FileSourceScanExec> will run on GPU

22/11/10 03:11:50 ERROR Executor: Exception in task 0.0 in stage 34.0 (TID 36)
java.lang.IllegalStateException: Column contains at least one value that is not in the required range
	at com.nvidia.spark.rapids.GpuCast$.$anonfun$castStringToBool$5(GpuCast.scala:1016)
	at com.nvidia.spark.rapids.GpuCast$.$anonfun$castStringToBool$5$adapted(GpuCast.scala:1014)
	at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:28)
	at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:26)
	at com.nvidia.spark.rapids.GpuCast$.withResource(GpuCast.scala:162)
	at com.nvidia.spark.rapids.GpuCast$.$anonfun$castStringToBool$4(GpuCast.scala:1014)

```

Maybe we should open another issue to support TRY_* mode  as this is only to fix current unit tests? I am not sure if there is a way to fall back only if it is try_cast ?  Please let us know how to go about this.

Reply:
On Spark-3.3, TryCast is separate Expression, so it falls back to CPU:

```
scala> spark.conf.set("spark.rapids.sql.enabled", "true")

scala>  spark.sql("select TRY_CAST(bools AS boolean) from boolsTable").show
22/11/10 05:38:30 WARN GpuOverrides: 
!Exec <CollectLimitExec> cannot run on GPU because the Exec CollectLimitExec has been disabled, and is disabled by default because Collect Limit replacement can be slower on the GPU, if huge number of rows in a batch it could help by limiting the number of rows transferred from GPU to CPU. Set spark.rapids.sql.exec.CollectLimitExec to true if you wish to enable it
  @Partitioning <SinglePartition$> could run on GPU
  !Exec <ProjectExec> cannot run on GPU because not all expressions can be replaced
    @Expression <Alias> cast(try_cast(bools#0 as boolean) as string) AS TRY_CAST(bools AS BOOLEAN)#12 could run on GPU
      @Expression <Cast> cast(try_cast(bools#0 as boolean) as string) could run on GPU
        ! <TryCast> try_cast(bools#0 as boolean) cannot run on GPU because GPU does not currently support the operator class org.apache.spark.sql.catalyst.expressions.TryCast
          @Expression <AttributeReference> bools#0 could run on GPU
    *Exec <FileSourceScanExec> will run on GPU

+--------------------------+
|TRY_CAST(bools AS BOOLEAN)|
+--------------------------+
|                      null|
|                      null|
+--------------------------+

```

Reply:
I don't care about supporting try_* until a customer asks for it. What I was concerned about is that with this change and our tests it looks like we were going to accidentally add in support for it without knowing. And it looks like that was true.  Just make sure that if the EvalMode is TRY we fall back to the CPU and that is good enough for me.  Fully supporting try mode is going to be very hard.